App/src/main/java/cc/softwarefactory/lokki/android/SettingsFragment.java

```diff
@@ -39,58 +45,63 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
-                    ServerAPI.setVisibility(context, false);
-                } else {
-                    MainApplication.visible = true;
-                    LocationService.start(context);
-                    ServerAPI.setVisibility(context, true);
-                }
-            } catch (Exception ex) {
-                ex.printStackTrace();
+        // Visibility
+        final TextView visibilityLabel = (TextView) rootView.findViewById(R.id.textView_visible);
+        setVisibilityInfoText(MainApplication.visible, visibilityLabel);
+
+        SwitchCompat toggle = (SwitchCompat) rootView.findViewById(R.id.switch_visibility);
+        toggle.setChecked(MainApplication.visible);
+        toggle.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
+            @Override
+            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+                Utils.setVisibility(isChecked, context);
+                setVisibilityInfoText(isChecked, visibilityLabel);
+            }
+        });
+
+
+        // Map modes
+        final String[] options = getResources().getStringArray(R.array.map_modes);
+        final TextView mapModeLabel = ((TextView)rootView.findViewById(R.id.textView_map_mode));
+        mapModeLabel.setText(options[getMapMode()]);
+
+        LinearLayout mapModeLayout = (LinearLayout) rootView.findViewById(R.id.settings_map_mode);
+        mapModeLayout.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                new AlertDialog.Builder(getActivity())
+                        .setTitle(R.string.map_mode)
+                        .setNegativeButton(R.string.cancel, null)
+                        .setSingleChoiceItems(options, getMapMode(), new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                PreferenceUtils.setValue(context, PreferenceUtils.KEY_SETTING_MAP_MODE, String.valueOf(which));
+                                MainApplication.mapType = which;
+                                mapModeLabel.setText(options[which]);
+                                dialog.dismiss();
+                            }
+                        })
+                        .show();
             }
+        });
 
-        } else if (tag.equals("map")) {
-            PreferenceUtils.setValue(context, PreferenceUtils.KEY_SETTING_MAP_MODE, String.valueOf(position));
-            MainApplication.mapType = position;
-        }
+        return rootView;
     }
 
-    @Override
-    public void onNothingSelected(AdapterView<?> parent) {
-
+    private int getMapMode () {
+        String map_mode = PreferenceUtils.getValue(context, PreferenceUtils.KEY_SETTING_MAP_MODE);
+        return (map_mode.equals("0") || map_mode.equals("1") || map_mode.equals("2")) ? Integer.valueOf(map_mode) : 0;
```

Comment:
This isn't perfect either but I think the following would be a better idiom here:

    Arrays.asList("0", "1", "2").contains(map_mode);

`map_mode` needs to be on camelCase, btw. Though this whole method is kinda in a wrong place here – I would have this method in PreferenceUtils or some other class that deals with saving and loading preferences that we care about.